File: scripts/cache_recipe_images.py
#!/usr/bin/env python3
"""Download recipe images locally and rewrite image URLs.

This script reads `recipes.json`, downloads each remote image into
`public/images/recipes/`, and rewrites the `imageUrl` field to point at
the cached copy. It skips empty/data URLs and preserves existing local
paths. Run it from the project root:

    python scripts/cache_recipe_images.py

Use `--help` for additional options.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import mimetypes
import re
import sys
from pathlib import Path
from typing import Dict, Iterable, Tuple
from urllib.error import URLError, HTTPError
from urllib.parse import urlparse
from urllib.request import Request, urlopen


DEFAULT_RECIPES_PATH = Path("recipes.json")
DEFAULT_OUTPUT_PATH = DEFAULT_RECIPES_PATH
DEFAULT_IMAGE_DIR = Path("public/images/recipes")
USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
    "AppleWebKit/537.36 (KHTML, like Gecko) "
    "Chrome/118.0 Safari/537.36"
)

SAFE_SLUG_RE = re.compile(r"[^a-z0-9]+")
VALID_EXTENSIONS = {".jpg", ".jpeg", ".png", ".gif", ".webp"}


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--recipes-file",
        type=Path,
        default=DEFAULT_RECIPES_PATH,
        help="Path to recipes JSON file (default: recipes.json)",
    )
    parser.add_argument(
        "--output-file",
        type=Path,
        default=DEFAULT_OUTPUT_PATH,
        help="Where to write the updated JSON (default: overwrite recipes file)",
    )
    parser.add_argument(
        "--image-dir",
        type=Path,
        default=DEFAULT_IMAGE_DIR,
        help="Directory to store cached images (default: public/images/recipes)",
    )
    parser.add_argument(
        "--skip-existing",
        action="store_true",
        help="Skip downloading if the target file already exists",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Do not download or modify files; just report planned actions",
    )
    return parser.parse_args()


def load_recipes(recipes_path: Path) -> Dict:
    with recipes_path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def slugify(title: str) -> str:
    slug = SAFE_SLUG_RE.sub("-", title.lower()).strip("-")
    if not slug:
        slug = f"recipe-{hashlib.sha1(title.encode('utf-8')).hexdigest()[:10]}"
    return slug[:60]


def guess_extension(url: str, content_type: str | None) -> str:
    parsed = urlparse(url)
    url_ext = Path(parsed.path).suffix.lower()
    if url_ext in VALID_EXTENSIONS:
        return url_ext

    if content_type:
        ext = mimetypes.guess_extension(content_type.split(";")[0].strip())
        if ext in VALID_EXTENSIONS:
            return ext

    return ".jpg"


def download_image(url: str) -> Tuple[bytes, str]:
    request = Request(url, headers={"User-Agent": USER_AGENT})
    with urlopen(request, timeout=30) as response:
        content = response.read()
        content_type = response.headers.get("Content-Type")
    return content, content_type


def ensure_directory(path: Path) -> None:
    path.mkdir(parents=True, exist_ok=True)


def _to_public_path(path: Path) -> str | None:
    try:
        rel = path.relative_to(Path("public"))
    except ValueError:
        return None
    return f"/{rel.as_posix()}"


def update_recipes(
    recipes: Dict,
    image_dir: Path,
    skip_existing: bool,
    dry_run: bool,
) -> bool:
    ensure_directory(image_dir)
    changed = False
    slug_counts: Dict[str, int] = {}

    for entry in recipes.get("recipes", []):
        image_url = (entry.get("imageUrl") or "").strip()
        title = entry.get("title", "").strip() or "recipe"

        if not image_url or image_url.startswith(("data:", "/")):
            continue

        slug = slugify(title)
        slug_counts.setdefault(slug, 0)
        slug_counts[slug] += 1
        final_slug = slug if slug_counts[slug] == 1 else f"{slug}-{slug_counts[slug]}"

        try:
            ext = guess_extension(image_url, None)
            target_file = image_dir / f"{final_slug}{ext}"

            if skip_existing and target_file.exists():
                new_url = _to_public_path(target_file)
                if new_url and entry.get("imageUrl") != new_url:
                    entry["imageUrl"] = new_url
                    changed = True
                continue

            if dry_run:
                print(f"[dry-run] Would download {image_url} -> {target_file}")
                continue

            content, content_type = download_image(image_url)
            ext = guess_extension(image_url, content_type)
            target_file = image_dir / f"{final_slug}{ext}"

            target_file.write_bytes(content)
            new_url = _to_public_path(target_file)
            if new_url:
                entry["imageUrl"] = new_url
            else:
                entry["imageUrl"] = str(target_file.as_posix())
            changed = True
            print(f"Downloaded {image_url} -> {target_file}")
        except HTTPError as error:
            print(f"[warning] HTTP error {error.code} for {image_url}", file=sys.stderr)
        except URLError as error:
            print(f"[warning] URL error for {image_url}: {error.reason}", file=sys.stderr)
        except Exception as exc:  # pylint: disable=broad-except
            print(f"[warning] Failed to cache {image_url}: {exc}", file=sys.stderr)

    return changed


def main() -> int:
    args = parse_args()

    recipes = load_recipes(args.recipes_file)
    changed = update_recipes(recipes, args.image_dir, args.skip_existing, args.dry_run)

    if args.dry_run:
        return 0

    if changed:
        with args.output_file.open("w", encoding="utf-8") as handle:
            json.dump(recipes, handle, ensure_ascii=False, indent=2)
            handle.write("\n")
        print(f"Wrote updated recipes to {args.output_file}")
    else:
        print("No changes made.")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/cacheRecipeImages.js
#!/usr/bin/env node
/**
 * Cache recipe images locally and rewrite their URLs.
 *
 * Usage:
 *   node scripts/cacheRecipeImages.js [options]
 *
 * Options:
 *   --recipes-file=<path>   Path to recipes JSON (default: recipes.json)
 *   --output-file=<path>    Write updated JSON here (default: overwrite recipes file)
 *   --image-dir=<path>      Directory for cached images (default: public/images/recipes)
 *   --skip-existing         Reuse existing files instead of redownloading
 *   --dry-run               Describe actions without changing files
 */

const fs = require('fs/promises');
const path = require('path');
const crypto = require('crypto');
const { request } = require('https');

const DEFAULT_RECIPES_PATH = path.resolve('recipes.json');
const DEFAULT_OUTPUT_PATH = DEFAULT_RECIPES_PATH;
const DEFAULT_IMAGE_DIR = path.resolve('public/images/recipes');
const USER_AGENT =
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) ' +
  'AppleWebKit/537.36 (KHTML, like Gecko) ' +
  'Chrome/118.0 Safari/537.36';
const VALID_EXTENSIONS = new Set(['.jpg', '.jpeg', '.png', '.gif', '.webp']);

function parseArgs(argv) {
  const options = {
    recipesFile: DEFAULT_RECIPES_PATH,
    outputFile: DEFAULT_OUTPUT_PATH,
    imageDir: DEFAULT_IMAGE_DIR,
    skipExisting: false,
    dryRun: false
  };

  argv.slice(2).forEach((arg) => {
    if (arg === '--skip-existing') {
      options.skipExisting = true;
    } else if (arg === '--dry-run') {
      options.dryRun = true;
    } else if (arg.startsWith('--recipes-file=')) {
      options.recipesFile = path.resolve(arg.split('=')[1]);
    } else if (arg.startsWith('--output-file=')) {
      options.outputFile = path.resolve(arg.split('=')[1]);
    } else if (arg.startsWith('--image-dir=')) {
      options.imageDir = path.resolve(arg.split('=')[1]);
    } else if (arg === '--help' || arg === '-h') {
      console.log(
        'Usage: node scripts/cacheRecipeImages.js [options]\n' +
          'Options:\n' +
          '  --recipes-file=<path>   Path to recipes JSON (default: recipes.json)\n' +
          '  --output-file=<path>    Write updated JSON here (default: overwrite recipes file)\n' +
          '  --image-dir=<path>      Directory for cached images (default: public/images/recipes)\n' +
          '  --skip-existing         Reuse existing files instead of redownloading\n' +
          '  --dry-run               Describe actions without changing files\n'
      );
      process.exit(0);
    }
  });

  return options;
}

async function loadRecipes(recipesPath) {
  const raw = await fs.readFile(recipesPath, 'utf8');
  return JSON.parse(raw);
}

function slugify(value) {
  const lower = value.toLowerCase();
  const slug = lower.replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, '');
  if (slug) {
    return slug.slice(0, 60);
  }
  const hash = crypto.createHash('sha1').update(value).digest('hex').slice(0, 10);
  return `recipe-${hash}`;
}

function guessExtension(url, contentType) {
  const parsed = new URL(url);
  const fromPath = path.extname(parsed.pathname).toLowerCase();
  if (VALID_EXTENSIONS.has(fromPath)) {
    return fromPath;
  }

  if (contentType) {
    const baseType = contentType.split(';')[0].trim();
    const lookup = {
      'image/jpeg': '.jpg',
      'image/jpg': '.jpg',
      'image/png': '.png',
      'image/gif': '.gif',
      'image/webp': '.webp'
    };
    if (lookup[baseType]) {
      return lookup[baseType];
    }
  }

  return '.jpg';
}

function downloadImage(url) {
  return new Promise((resolve, reject) => {
    const req = request(
      url,
      {
        headers: {
          'User-Agent': USER_AGENT
        }
      },
      (res) => {
        if (res.statusCode && res.statusCode >= 400) {
          reject(new Error(`HTTP ${res.statusCode}`));
          res.resume();
          return;
        }
        const data = [];
        res.on('data', (chunk) => data.push(chunk));
        res.on('end', () => {
          resolve({
            buffer: Buffer.concat(data),
            contentType: res.headers['content-type'] || null
          });
        });
      }
    );

    req.on('error', reject);
    req.setTimeout(30000, () => {
      req.destroy(new Error('Request timeout'));
    });
    req.end();
  });
}

async function ensureDir(dirPath) {
  await fs.mkdir(dirPath, { recursive: true });
}

function toPublicPath(filePath) {
  const publicRoot = path.resolve('public');
  const absolute = path.resolve(filePath);
  if (!absolute.startsWith(publicRoot)) {
    return null;
  }
  return `/${path.relative(publicRoot, absolute).replace(/\\/g, '/')}`;
}

async function cacheImages(recipes, options) {
  await ensureDir(options.imageDir);

  let changed = false;
  const slugCounts = new Map();

  for (const recipe of recipes.recipes || []) {
    const originalUrl = (recipe.imageUrl || '').trim();
    const title = (recipe.title || 'recipe').trim();

    if (!originalUrl || originalUrl.startsWith('data:') || originalUrl.startsWith('/')) {
      continue;
    }

    const baseSlug = slugify(title);
    const count = (slugCounts.get(baseSlug) || 0) + 1;
    slugCounts.set(baseSlug, count);
    const slug = count === 1 ? baseSlug : `${baseSlug}-${count}`;

    const preliminaryExt = guessExtension(originalUrl, null);
    const targetPath = path.join(options.imageDir, `${slug}${preliminaryExt}`);
    const publicPath = toPublicPath(targetPath);

    if (options.skipExisting && publicPath && (await fileExists(targetPath))) {
      if (recipe.imageUrl !== publicPath) {
        recipe.imageUrl = publicPath;
        changed = true;
      }
      continue;
    }

    if (options.dryRun) {
      console.log(`[dry-run] Would download ${originalUrl} -> ${targetPath}`);
      continue;
    }

    try {
      const { buffer, contentType } = await downloadImage(originalUrl);
      const finalExt = guessExtension(originalUrl, contentType);
      const finalPath = path.join(options.imageDir, `${slug}${finalExt}`);
      await fs.writeFile(finalPath, buffer);
      const finalPublic = toPublicPath(finalPath);
      if (finalPublic) {
        recipe.imageUrl = finalPublic;
      } else {
        recipe.imageUrl = finalPath;
      }
      changed = true;
      console.log(`Downloaded ${originalUrl} -> ${finalPath}`);
    } catch (error) {
      console.warn(`[warning] Failed to cache ${originalUrl}: ${error.message}`);
    }
  }

  return changed;
}

async function fileExists(filePath) {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

async function main() {
  const options = parseArgs(process.argv);

  try {
    const recipes = await loadRecipes(options.recipesFile);
    const changed = await cacheImages(recipes, options);

    if (options.dryRun) {
      return 0;
    }

    if (changed) {
      const data = `${JSON.stringify(recipes, null, 2)}\n`;
      await fs.writeFile(options.outputFile, data, 'utf8');
      console.log(`Wrote updated recipes to ${options.outputFile}`);
    } else {
      console.log('No changes made.');
    }
  } catch (error) {
    console.error(`[error] ${error.message}`);
    return 1;
  }

  return 0;
}

main().then(
  (code) => {
    process.exit(code);
  },
  (error) => {
    console.error('[error]', error);
    process.exit(1);
  }
);
